codec/reflectcodec/type_codec.go

```diff
@@ -31,69 +24,25 @@ var (
 	errExtraSpace   = errors.New("trailing buffer space")
 )
 
-// Codec handles marshaling and unmarshaling of structs
-type codec struct {
-	lock         sync.RWMutex
-	tagName      string
-	maxSliceLen  int
-	nextTypeID   uint32
-	typeIDToType map[uint32]reflect.Type
-	typeToTypeID map[reflect.Type]uint32
-
-	fieldLock sync.Mutex
-	// Key: a struct type
-	// Value: Slice where each element is index in the struct type
-	// of a field that is serialized/deserialized
-	// e.g. Foo --> [1,5,8] means Foo.Field(1), etc. are to be serialized/deserialized
-	// We assume this cache is pretty small (a few hundred keys at most)
-	// and doesn't take up much memory
-	serializedFieldIndices map[reflect.Type][]int
+type TypeCodec interface {
+	UnpackPrefix(*wrappers.Packer, reflect.Type) (reflect.Value, error)
+	PackPrefix(*wrappers.Packer, reflect.Type) error
```

Comment:
// PackPrefix packs the prefix for the given type into the given packer.
// This identifies the bytes that follow, which are the byte representation of an interface, as having the given concrete type.
// When deserializing the bytes, the prefix specifies which concrete type to deserialize into.